### ISSUE TITLE
Services with protection features fail to start if root fs is read-only

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#14324]
A whole bunch of services fail to start if / is not remounted rw correctly. Most (all?) would start and actually run correctly without any issue, if the protection settings weren't there. For general resiliency, I think we should handle this more gracefully. In general, the protection settings should be additive, and not cause things to break completely if something goes wrong. If the service fails because it cannot write stuff then it should fail on its own. E.g. PrivateTmp=yes (for /var/tmp) and ReadWritePaths=/etc cause issues.

```console
$ systemctl --failed
  UNIT                       LOAD   ACTIVE SUB    DESCRIPTION                  
● systemd-remount-fs.service loaded failed failed Remount Root and Kernel File  ← OK
● systemd-logind.service     loaded failed failed Login Service
● dbus.socket                loaded failed failed D-Bus System Message Bus Sock
● dbus-broker.service        loaded failed failed D-Bus System Message Bus     
```
`systemd-remount-fs.service` fails correctly, but the other ones shouldn't.

(See also #2339.)

**Comment by poettering** [comment]
Do you know which protection setting is causing this?

**Comment by keszybz** [comment]
PrivateTmp=yes (for /var/tmp) and ReadWritePaths=/etc.

https://github.com/keszybz/systemd/commit/8c6b34ad3083caa7a0a6c5c0759e416c82ef43fd is the easy fix for the second one.

**Comment by poettering** [comment]
is it possible /var/tmp and /etc are simply missing on the root disk? i.e. so that we cannot create them to overmount them?

**Comment by boucman** [comment]
FTR I rencently compiled systemd without systemd-tmpfiles and I had exactly the same symptoms becaus (indeed) nobody created /var/tmp. This caused https://github.com/systemd/systemd/pull/14207

**Comment by poettering** [comment]
@keszybz or to say this differently: for read-only images it's assumed you created all necessary mount points when preparing the image, and that includes /var/tmp and /tmp.

**Comment by keszybz** [comment]
> is it possible /var/tmp and /etc are simply missing on the root disk?

No. The problem is that systemd needs to create `/var/tmp/systemd-private-a1b54d219c024b5fb9dc0d27871e3046-foo` for the private /var/tmp/ directory of the service. Obviously, this fails.

**Comment by dustwolf** [comment]
I'm having a similar issue while trying to move an Ubuntu installation to use an overlay root.

The `systemd-remount-fs.service` part is easily fixed by using Casper's `/etc/fstab`:
```
overlay / overlay rw 0 0
tmpfs /tmp tmpfs nosuid,nodev 0 0
```
However the issue with D-Bus and everything that uses D-Bus remains. In my case I am using a persistent upper overlay (ext4), so I can see that systemd can create the files in /var/tmp/ without a problem.

I suspect that a different directory is missing (the system was created using rsync), but I do not know which. I will see if I can reverse-engineer this from Casper.

**Comment by wjt** [comment]
I recently accidentally filled my filesystem as root, and attempted to reboot. logind and a bunch of other services failed to come up. Since logind didn't come up, GDM didn't come up, and I was unable to log in to delete some files; fortunately I know how to resort to `systemd.debug_shell=1`.

> > is it possible /var/tmp and /etc are simply missing on the root disk?
> 
> No. The problem is that systemd needs to create `/var/tmp/systemd-private-a1b54d219c024b5fb9dc0d27871e3046-foo` for the private /var/tmp/ directory of the service. Obviously, this fails.

I'm pretty sure this explains what I saw: most of the services which failed to start failed with:

```
systemd-logind.service: Failed to run 'start-pre' task: No space left on device
systemd-logind.service: Failed with result 'resources'.
```

and they all have `PrivateTmp=yes`. It would have been kinder for logind in particular to degrade gracefully to a less-sandboxed state so that I could log in and fix the situation.

**Comment by keszybz** [comment]
Fixed in 56a13a495ce82d431bc68e0939b76373c879bed3.
